File: 42-trapping-rain-water/42-trapping-rain-water.py
class Solution:
    def trap(self, h: List[int]) -> int:
      N = len(h)
      if(N<=2):
        return 0
      l = [0]*N
      r = [0]*N
      l[0] = h[0]
      r[N-1] = h[N-1]
      for i in range(1, N):
        l[i] = max(l[i-1], h[i])
        r[N-i-1] = max(r[N-i], h[N-i-1])
      ret = 0;
      for i in range(N):
        ret += min(l[i], r[i]) - h[i]
      return ret
        

File: 42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
      int n = height.size();
      int ans = 0;
      if(n<3)
        return 0;
      
      vector<int> l(n, 0);
      vector<int> r(n, 0);
      
      l[0]    =  height[0];
      r[n-1]  =  height[n-1];
      
      for(int i=1; i<n; i++)
        l[i] = max(l[i-1], height[i]); 
      for(int i=1; i<n; i++)
        r[n-i-1] = max(r[n-i], height[n-i-1]); 
      for(int i=0; i<n; i++)
        ans += min(r[i], l[i]) - height[i];
      return ans;
    }
};